addon/-private/system/relationships/state/implicit.js

```diff
@@ -0,0 +1,200 @@
+/* global heimdall */
+import OrderedSet from '../../ordered-set';
+import Relationship from './relationship';
+
+const {
+  addCanonicalRecord,
+  addCanonicalRecords,
+  addRecord,
+  addInternalModels,
+  clear,
+  flushCanonical,
+  newRelationship,
+  removeCanonicalRecord,
+  removeCanonicalRecordFromOwn,
+  removeCanonicalRecords,
+  removeRecord,
+  removeRecordFromOwn,
+  removeInternalModels,
+  updateRecordsFromAdapter
+} = heimdall.registerMonitor('system.relationships.state.relationship',
+  'addCanonicalRecord',
+  'addCanonicalRecords',
+  'addRecord',
+  'addInternalModels',
+  'clear',
+  'flushCanonical',
+  'newRelationship',
+  'removeCanonicalRecord',
+  'removeCanonicalRecordFromOwn',
+  'removeCanonicalRecords',
+  'removeRecord',
+  'removeRecordFromOwn',
+  'removeInternalModels',
+  'updateRecordsFromAdapter'
+);
+
+export default class ImplicitRelationship extends Relationship {
+  constructor(store, internalModel, inverseKey, relationshipMeta) {
+    super(store, internalModel, inverseKey, relationshipMeta);
+    this.kind = 'implicit';
+    heimdall.increment(newRelationship);
+
+    this.members = new OrderedSet();
+    this.canonicalMembers = new OrderedSet();
+  }
+
+  destroy() {
+    if (!this.inverseKey) { return; }
+
+    let allMembers =
```

Comment:
(also fwiw this is the legacy path that everything used to do, now only implicit does it, I haven't spent much time optimizing implicit's code yet)